pkg/workspace/assert.go

```diff
@@ -0,0 +1,162 @@
+// Copyright 2022 VMware, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package workspace
+
+import (
+	"fmt"
+	"github.com/k14s/starlark-go/starlark"
+	"github.com/vmware-tanzu/carvel-ytt/pkg/filepos"
+	"github.com/vmware-tanzu/carvel-ytt/pkg/template"
+	"github.com/vmware-tanzu/carvel-ytt/pkg/yamlmeta"
+	"github.com/vmware-tanzu/carvel-ytt/pkg/yamltemplate"
+)
+
+// Declare @assert/... annotation names
+const (
+	AnnotationAssertValidate template.AnnotationName = "assert/validate"
+)
+
+// ProcessAndRunValidations takes a root Node and traverses the tree checking for assert annotations.
+// Validations are processed and executed using the value of the annotated node as the parameter to the assertions.
+//
+// When the assertions have violations, the errors are collected and stored in the checker.
+// Otherwise, returns nil.
+func ProcessAndRunValidations(n yamlmeta.Node, checker yamlmeta.Visitor) error {
+	if n == nil {
+		return nil
+	}
+
+	err := yamlmeta.Walk(n, checker)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
```

Comment:
The function requires the caller to construct the checker.

In a previous revision, it _took on_ that responsibility... what motivated relocating that responsibility to the caller(s)?

I think passing in the name of the scenario/task (which we set as the name of the thread) — in this case "assert-data-values" — seems to fit, given that we want to run validations in a number of places. 👍🏻 

When I naively think of what I'd expect the signature of this function to be, I assume:
- the Node(tree)
- some configuration about the validation process itself (so far, we've identified the need to give it a name... in the future, we might provide _other_ options. 🤷🏻 )
- and it would return the slice of violations.

Perhaps we're want for a value object that holds the violations that has that `HasViolations()` method on it? Perhaps we extract that out of the `assertChecker` — it would then _produce_ such a value rather than _be_ that value... 🤔

Reply:
What motivated this change was that I wanted `ProcessAndRunValidations()` to be able to accept different kinds of checkers. Right now there is just one Validation annotation, but I was thinking that if we needed to process a subset of Validation annotations, then accepting a Visitor interface would be the way to do it.

Reply:
Seems like I may be assuming too much about the future in this case, I agree that the responsibility should be with `ProcessAndRunValidations()` to construct the checker.